python-package/xgboost/training.py

```diff
@@ -514,8 +514,6 @@ def cv(
         params = dict(params)
         if 'eval_metric' in params:
             params['eval_metric'] = _metrics
-    else:
-        params = dict((k, v) for k, v in params.items())
```

Comment:
`flake8-comprehensions` warned about the use of `dict()` here, but when I looked more closely, it seems like this line isn't necessary at all.

It's been in `xgboost` for 6+ years (since #1142), so maybe something has changed since then.

But if the code enters this point, it's already guaranteed that `params` is a dictionary.

https://github.com/dmlc/xgboost/blob/67752e39679ec3a5a9e3a4a4b93967931e4b7d62/python-package/xgboost/training.py#L512-L518

And for something that is already a dictionary, I can't think of a situation where re-iterating over all the contents like this would produce a different dictionary.

```shell
params = {"a": 1, "b": [2, 3], "c": None}

new_params = dict((k, v) for k, v in params.items())

params == new_params
# True
```

Reply:
Thank you for catching that!

Reply:
@trivialfis I found that this change was actually to blame for the failure of `tests/python/test_early_stopping.py`, not the one from this comment thread: https://github.com/dmlc/xgboost/pull/8546#discussion_r1042219019.

That `dict((k, v) ...` code was serving a second purpose...it ensured that the rest of `xgboost.cv()` worked on *a copy* of `params`, so calls like `params.pop()` wouldn't have side-effects on the passed-in dictionary.

I fixed that in https://github.com/dmlc/xgboost/pull/8546/commits/849ea9221367fcb5ce8175befb523801b9f0d66a.

------

There is one remaining test Python test failure (`python/test_with_sklearn.py::test_pandas_input`) on macOS ([build link](https://github.com/dmlc/xgboost/actions/runs/3654519571/jobs/6175015371)) and one on Windows ([build link](https://github.com/dmlc/xgboost/actions/runs/3654519571/jobs/6175016596)).

```text
assert isinstance(
            clf_isotonic.calibrated_classifiers_[0].base_estimator, xgb.XGBClassifier
        )
E       AttributeError: '_CalibratedClassifier' object has no attribute 'base_estimator'
```

I think that might not be related to this PR's changes, given that I also see it on this other PR which only touches the R package: #8565 ([build link](https://github.com/dmlc/xgboost/actions/runs/3652941683/jobs/6171873398)).

I wonder if that's related to the newest release of `scikit-learn` supporting pandas as an input and output (https://github.com/scikit-learn/scikit-learn/issues/5523). 

I noticed, for example, that the failing Mac build here is using `scikit-learn=1.2.0`, which was just released today ([PyPI link](https://pypi.org/project/scikit-learn/#history)).

Reply:
Hopefully #8580 will fix this.